8263421: Module image file is opened twice during VM startup

hseigel opened the pull request:
Please review this fix for JDK-8263421.  The fix stores the JImageFile* in a static that gets initialized when lookup_vm_options() opens the module image file.  The static JimageFIle* is then used by setup_bootstrap_search_path_impl() when creating the ClassPathImageEntry for the module image file, avoiding having to re-open the module image file.

This fix moved the creation of the ClassPathImageEntry for the module image file from create_class_path_entry() to setup_bootstrap_search_path_impl() because create_class_path_entry() may try to do this multiple times, but the ClassPathImageEntry for the module image file is only actually created once.

The fix also removes some unused code related to module image file processing.

The changes were tested with Mach5 tiers 1 and 2 on Linux, Mac OS, and Windows, and Mach5 tiers 3-
5 on Linux x64.

Thanks, Harold

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263421](https://bugs.openjdk.java.net/browse/JDK-8263421): Module image file is opened twice during VM startup


### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 5341b1ededfd5f509605dbe73d106f5e3034a819
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3591/head:pull/3591` \
`$ git checkout pull/3591`

Update a local copy of the PR: \
`$ git checkout pull/3591` \
`$ git pull https://git.openjdk.java.net/jdk pull/3591/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3591`

View PR using the GUI difftool: \
`$ git pr show -t 3591`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3591.diff">https://git.openjdk.java.net/jdk/pull/3591.diff</a>

</details>

iklam commented on src/hotspot/share/classfile/classLoader.cpp:
```diff
@@ -386,15 +390,15 @@ ClassFileStream* ClassPathImageEntry::open_stream(Thread* current, const char* n
 ClassFileStream* ClassPathImageEntry::open_stream_for_loader(Thread* current, const char* name, ClassLoaderData* loader_data) {
   jlong size;
-  JImageLocationRef location = (*JImageFindResource)(_jimage, "", get_jimage_version_string(), name, &size);
+  JImageLocationRef location = (*JImageFindResource)(jimage(), "", get_jimage_version_string(), name, &size);
```
There's a subtle change:

* before: `_jimage` could never be NULL, since it's a private member of this `ClassPathImageEntry`
* after: `jimage()` could theoretically be NULL, if `ClassPathImageEntry::close_jimage()` is called, which happens only when:
  * `~ClassPathImageEntry()` is called. But this will never happen because we never delete `ClassPathImageEntry`;
  * `ClassLoader::close_jrt_image()` is called. This happens only on Linux inside `os::abort()`

When we call `jimage()` and expect it to be non-null, maybe we should use a new function `jimage_non_null()` like this:

```
JImageFile* ClassPathImageEntry::jimage_non_null() {
  assert(ClassLoader::has_jrt_entry(), "must be");
  assert(jimage() != NULL, "should have been opened by ClassLoader::lookup_vm_options "
                            "and remained throughout normal JVM lifetime");
  return jimage();
}
```

and we should change `~ClassPathImageEntry()` to assert that it's never called?

iklam reviewed (changes requested):
Looks good overall, but I think we should add some asserts to ensure safety.

hseigel commented:
Thanks Ioi for looking at this change!  I added jimage_non_null() and modified ~ClassPathImageEntry() as you suggested.  Could you take at these updated changes?

Thanks, Harold

iklam reviewed (approved):
LGTM

dholmes-ora commented on src/hotspot/share/classfile/classLoader.cpp:
```diff
@@ -105,10 +105,11 @@ int ClassLoader::_libzip_loaded = 0;
-static JImageResourceIterator_t        JImageResourceIterator = NULL;
+
+// JimageFile pointer, or null if exploded JDK build.
+static JImageFile*                     JImage_File            = NULL;
```
Naming style looks odd here - JImage_file ? _jimage_file ?

dholmes-ora commented on src/hotspot/share/classfile/classLoader.cpp:
```diff
@@ -343,16 +344,30 @@ void ClassPathZipEntry::contents_do(void f(const char* name, void* context), voi
+  return jimage();
+}
+
+void ClassPathImageEntry::set_jimage(JImageFile* jimage) {
+  JImage_File = jimage;
+}
```
This setter seems overkill, especially when it is only used to clear the field, and there is no setter used when giving the field a real value.

dholmes-ora reviewed (approved):
Hi Harold,

Generally this looks okay. A couple of minor nits.

Thanks,
David

hseigel commented:
Thanks David for reviewing this.  I removed the set_jimage() method and renamed static variable JImage_File to JImage_file.
Harold

hseigel commented:
Thanks Ioi and David for the reviews!

/integrate
